File: C/3.5-prompt/Par/sgemv_openmp.c
#include <omp.h>

void sgemv(int m, int n, float alpha, float *a, float *x, float beta, float *y)
{
    #pragma omp parallel for
    for (int i = 0; i < m; i++)
    {
        float sum = 0.0;
        for (int j = 0; j < n; j++)
        {
            sum += a[i * n + j] * x[j];
        }
        y[i] = alpha * sum + beta * y[i];
    }
}

File: Fortran/3.5/sgemv_openmp.f90
subroutine gemv_parallel(n, A, x, y)
    implicit none
    integer, intent(in) :: n
    real, intent(in) :: A(n, n)
    real, intent(in) :: x(n)
    real, intent(out) :: y(n)
    integer :: i, j
    real :: sum

    !$OMP PARALLEL DO
    do i = 1, n
        sum = 0.0
        !$OMP DO
        do j = 1, n
            sum = sum + A(i, j) * x(j)
        end do
        !$OMP END DO
        y(i) = sum
    end do
    !$OMP END PARALLEL DO

end subroutine gemv_parallel
